Brackets Health Report implementation

prafulVaishnav opened the pull request:
Initial Brackets Health Report implementation - see http://blog.brackets.io/2015/03/27/introducing-brackets-health-report/ for more information.
- If user has opt in for the tracking we will send the data for every 24 hours to the server.
- There is menu item under Help->HealthData Statistics to preview the health data that will be sending to the server if user has opt in for the tracking as well as the option to opt out of tracking. Same menu item is used to show the dialog box which has option to opt in for tracking.

For initial implementation, we are logging the data mentioned in initial-data-proposal https://github.com/adobe/brackets/wiki/Health-Data#initial-data-proposal

prafulVaishnav committed:
Implementation for Notification Dialog at the startup where user can set preference to track the health data.
Implementation for Preview Dialog where user can see the preview of the data that will be sending to the server.

prafulVaishnav committed:
Implementation of Modules for Client Health Implementation

prafulVaishnav committed:
Added some comments in the file. Implementation to get user Installed extension for health data.

prafulVaishnav committed:
Only user extensions that are present in registry are being tracked.
Added registryObject which will contain the registry object when call to downloadRegistry is made.
While sending data to the server or previewing the health data, if registry
object is not present, then call to downloadRegistry to get the object.

prafulVaishnav committed:
Change in the Dialog Boxes..

prafulVaishnav committed:
Modification in modules..

prafulVaishnav committed:
Added Unit tests for Health Data.. Modifications in the module implementations.

prafulVaishnav committed:
Add comments..

prafulVaishnav committed:
Deleting duplicate string.

prafulVaishnav committed:
Merge branch 'prafulVaishnav/ClientHealthData' of https://github.com/adobe/brackets into prafulVaishnav/ClientHealthData

prafulVaishnav commented:
@prksingh @abose Can you please review this PR?

prafulVaishnav committed:
Correcting jshint errors.

peterflynn commented:
Fyi - [I wrote up a blog post](http://blog.brackets.io/2015/03/27/introducing-brackets-health-report/) talking about the goals of this PR, and sharing other Brackets usage statistics that we already have available.

I'll edit the PR description above too, so anyone who sees this and is curious what it's about knows where to look.  And on Monday (when more people are actually paying attention :-) we'll post to brackets-dev/etc. pointing to the blog post, to make sure it doesn't go unnoticed.

prafulVaishnav commented:
@peterflynn Thanks a lot Peter..

MarcelGerber commented on src/extensibility/ExtensionManager.js:
```diff
@@ -260,6 +280,18 @@ define(function (require, exports, module) {
+                    userExtensions[id] = {};
+                }
+                
+                userExtensions[id].installInfo = {
```
Can't we also set the data directly on `userExtensions[id]` rather than `userExtensions[id].installInfo`?
Is there a plan to ever have more properties than just `installInfo`?

prafulVaishnav replied:
@MarcelGerber Currently also we have other properties like registryInfo and most probably others can be added.

MarcelGerber commented on src/healthData/HealthDataManager.js:
```diff
@@ -0,0 +1,180 @@
+        PreferencesManager = require("preferences/PreferencesManager"),
+        HealthDataUtils = require("healthData/HealthDataUtils"),
+        uuid = require("thirdparty/uuid"),
+        UrlParams = require("utils/UrlParams").UrlParams;
```
nit: Indent all these requires to the same level.

prafulVaishnav replied:
@MarcelGerber Thanks.. Will do.

MarcelGerber commented on src/healthData/HealthDataManager.js:
```diff
@@ -0,0 +1,180 @@
+        var result = new $.Deferred(),
+            oneTimeHealthData = {};
+
+        var guid = PreferencesManager.getViewState("GUID");
```
GUID is actually Microsoft's terminology for an UUID ([source](http://stackoverflow.com/a/246935/3455922)), so I guess we should rather call it UUID.

prafulVaishnav replied:
@MarcelGerber Yes. You are right. We are generating UUID in our case. Will make change.

MarcelGerber commented on src/healthData/HealthDataManager.js:
```diff
@@ -0,0 +1,180 @@
+    params.parse();
+
+    /**
+     * Get the health data which will be send to the server. Initially it is only one time data.
```
nit: `will be sent` (with t instead of d)

prafulVaishnav replied:
@MarcelGerber Thanks.. Will do.

MarcelGerber commented on src/healthData/HealthDataManager.js:
```diff
@@ -0,0 +1,180 @@
+    function sendHealthDataToServer() {
+        var result = new $.Deferred();
+        
+        getHealthData().done(function (jsonData) {
```
It's not really `jsonData`, it's just a normal JS object.
I'd expect a JSON string being stored in a variable called `jsonData`.

Better just call it `data` or `healthData`.

prafulVaishnav replied:
@MarcelGerber name jsonData is used just to clarify the structure of healthdata. Integration of server and client has not been done yet so I have just left this part to complete after integration

MarcelGerber commented on src/healthData/HealthDataManager.js:
```diff
@@ -0,0 +1,180 @@
+
+            if (!lastTimeSend) {
+                var randomTime = Math.floor(Math.random() * ONE_DAY);
+                lastTimeSend = currentTime + randomTime;
```
Why do we add a random time here?

prafulVaishnav replied:
@MarcelGerber To randomize the time when the data will be sent to the server. It is distributed within 24 hours.

MarcelGerber replied:
It won't work though, it will actually upload the data within 24 to 48 hours.

Just take a look at the `setTimeout` in the `else` case. With `lastTimeSend` being `currentTime + randomTime`, it says `(currentTime + randomTime) + ONE_DAY - currentTime`, which simplifies to `randomTime + ONE_DAY` -- always something in the range of 24-48h.

Also, I still don't quite get why you don't immediately send the data if it wasn't ever sent before (or maybe with a little timeout around 5 minutes, to "warm Brackets up").

prafulVaishnav replied:
@MarcelGerber lastTimeSend will be (currentTime + randomTime) only at first launch of Brackets when pref "var lastTimeSend = **PreferencesManager.getViewState("lastTimeSendHealthData")**" is not set. So for the first time it will be in the range of 24-48 hours. After that, it will be always 24 hour **timeoutVar = setTimeout(_checkHealthDataExport, ONE_DAY);**

Purpose of adding random time is to prevent enormous request to the server  It is possible as soon as new version of Brackets will be available users will update it and time of sending data to the server will possibly collide for many users. 

We need to have some minimum time period for sending data to the server which in our case is 24 hour.

peterflynn replied:
@MarcelGerber We don't want to immediately send data on first launch because the user needs time to opt out if they want to.

@prafulVaishnav Remember preferences persist on upgrades, so when the big upgrade spike happens the vast majority of users who are upgrading won't hit this codepath anyway.  It's only really an issue if some sort of viral PR thing happened, with many _brand new_ users coming in at once -- and even then I'd expect new-user first launches to be distributed in time enough that randomization may not be needed.

Since it adds complexity, we should think about whether this wrinkle in the logic is really needed.  If you still feel like it's important, please add a comment though explaining that the purpose is to distribute server load (and pointing out that the _minimum_ time before the first ping is 24 hours, due to the `+ ONE_DAY` in the `else` case below).

prafulVaishnav replied:
@peterflynn Initially we decided to have random time to distribute load on server. But @abose @prksingh have tested the load on server and it can easily handle requests. So no randomization needed now.

MarcelGerber commented on src/healthData/HealthDataNotification.js:
```diff
@@ -0,0 +1,93 @@
+        PreferencesManager           = require("preferences/PreferencesManager"),
+        Strings                      = require("strings"),
+        Dialogs                      = require("widgets/Dialogs"),
+        HealthDataNotificationDialog = require("text!htmlContent/healthdata-notification-dialog.html"),
```
Usually, text file are the last ones in a require section.

prafulVaishnav replied:
@MarcelGerber Thanks. Will do..

MarcelGerber commented on src/healthData/HealthDataNotification.js:
```diff
@@ -0,0 +1,93 @@
+        });
+
+        Dialogs.showModalDialogUsingTemplate($template).done(function (id) {
+            PreferencesManager.setViewState("healthDataNotification", (new Date()).getTime());
```
Why do we save a timestamp over here? Can't we just use a boolean?

prafulVaishnav replied:
@MarcelGerber If user opt out for health data tracking and we decide to show the notification for opt-in after some time period let say 1 week, then it will be used there.

peterflynn replied:
I'm also not sure I understand what this date is needed for.  Could you explain the workflow in a little more detail?  Also, are you saying the code in this PR uses the date for a specific reason, or is just just a hypothetical future use case?

peterflynn replied:
I also think the name of this preference is confusing.  If it's a date, we should call it something like `"healthDataNotificationLastShown"`.  If it's a boolean, maybe `"healthDataNotificationShown"`.

prafulVaishnav replied:
@peterflynn @MarcelGerber Currently not using it atleast for this release. It was future use case, so will remove it for now.

MarcelGerber commented on src/healthData/HealthDataNotification.js:
```diff
@@ -0,0 +1,93 @@
+            PreferencesManager.setViewState("healthDataNotification", (new Date()).getTime());
+     
+            if (id === "save") {
+                if (hdPref !== newHDPref) {
```
I _think_ we can gather the checkbox state here rather than listening for change events.

prafulVaishnav replied:
@MarcelGerber At this point we don't have dialog box so not possible to get value of an element

MarcelGerber replied:
```JS
console.log($template.find("input").prop("checked"));
```
worked for me at this point.

peterflynn replied:
Yeah, I agree this could be simplified.  Also, the check for whether the value has changed probably isn't needed (there's no harm in setting the pref either way, right? -- I hope not anyway, since the prefs layer sends extra change notifications all the time)

MarcelGerber commented on src/healthData/HealthDataPreview.js:
```diff
@@ -0,0 +1,75 @@
+    /** 
+     * Show dialog for previewing the data send to server for Health Data
+     */
+    function previewHealthDataFile() {
```
This function name is misleading as it's not actually a file.
I'd simply go with `previewHealthData`.

prafulVaishnav replied:
@MarcelGerber Yeah Thanks for catching it. Actually I was logging some events that need to persist in the file so was saving the healthdata in the file and previewing the file. But for first release we decided to minimize the event logging that need not to be persist. I missed to change the name.

MarcelGerber commented on src/healthData/HealthDataPreview.js:
```diff
@@ -0,0 +1,75 @@
+        HealthDataManager.getHealthData().done(function (healthDataObject) {
+            var content = JSON.stringify(healthDataObject, null, 4);
+            content = content.replace(/ /g, "&nbsp;");
+            content = content.replace(/(?:\r\n|\r|\n)/g, "<br />");
```
You can use a [`pre` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/pre) so you don't have to deal with replacing these characters.
You should still escape some characters, but a simple method like [`_.escape`](https://lodash.com/docs#escape) should suffice.

prafulVaishnav replied:
@MarcelGerber I think handling space and newline will suffice here instead of using pre as we will not be having much-formatted text.

peterflynn replied:
We should still call `_.escape()` first though -- otherwise strings containing "<" could break the markup.

MarcelGerber commented on src/healthData/HealthDataUtils.js:
```diff
@@ -0,0 +1,112 @@
+            "5.1" : "Windows XP",
+            "5.2" : "Windows XP | Windows Server 2003 | Windows Home Server",
+            "6.0" : "Windows Vista | Windows Server 2008",
+            "6.1" : "Windows 7 | Widnows Server 2008 | Widnows Home Server 2011",
```
nit: `Widnows`

prafulVaishnav replied:
@MarcelGerber Thanks.. Will do.

MarcelGerber commented on src/healthData/HealthDataUtils.js:
```diff
@@ -0,0 +1,112 @@
+        };
+        
+        var windowRegex = /Windows NT (\d+\.\d+)/i,
+            macRegex    = /Macintosh;\s([\w\W]+?)(\)|;)[\w\W]+/i,
```
Instead of using `[\w\W]`, use `.` and enable the jslint flag `regexp: true` (as we did with multiple other files).

prafulVaishnav replied:
@MarcelGerber Yes, both are same things. But don't know how it became by habit to write this way :)

MarcelGerber commented on src/healthData/HealthDataUtils.js:
```diff
@@ -0,0 +1,112 @@
+        } else if (/Mac/.test(userAgent)) {
+            osVersion = macRegex.exec(userAgent)[1] || osVersion;
+        } else if (/X11/.test(userAgent)) {
+            osVersion = linuxRegex.exec(userAgent) || osVersion;
```
This will give you an array.

prafulVaishnav replied:
@MarcelGerber Thanks for catching it..

MarcelGerber commented on src/healthData/HealthDataUtils.js:
```diff
@@ -0,0 +1,112 @@
+        
+        if (Object.keys(registryExtensions).length === 0) {
+            ExtensionManager.downloadRegistry().done(function () {
+                return result.resolve(getUserExtensionsInRegistry(ExtensionManager.registryObject, userExtensions));
```
Remove the `return` statement in all these handlers. It's a no-op.

prafulVaishnav replied:
@MarcelGerber Thanks.. Will do..

MarcelGerber commented on src/htmlContent/healthdata-preview-dialog.html:
```diff
@@ -0,0 +1,23 @@
+    </div>
+    <div class="modal-body">
+	    <div class="content">
+	        <p class="dialog-message preview-content">{{{content}}}</p>
```
As mentioned above: Rather use a `pre` element.

le717 commented on src/healthData/HealthDataUtils.js:
```diff
@@ -0,0 +1,112 @@
+        var winOSStrings = {
+            "5.0" : "Windows 2000",
+            "5.1" : "Windows XP",
+            "5.2" : "Windows XP | Windows Server 2003 | Windows Home Server",
```
Should Windows XP (much less 2000) even be listed? XP support was dropped [last April](http://blog.brackets.io/2014/04/18/windows-xp-support/), and AFAIK 2000 has never been supported.

prafulVaishnav replied:
@le717 Yeah you are right. I missed it while writing.. Thanks for catching it.

MarcelGerber replied:
I don't think we should catch it just yet -- I'm pretty sure there are still some using Brackets on XP, and it's way more useful to have the actual OS name on the server rather than just an User Agent string.
Also, it won't hurt to leave XP in for now.

prafulVaishnav replied:
@MarcelGerber @le717 I  am kinda unaware of the minimum supported platforms where user can install Brackets. XP support is not there but if there is Brackets running on XP then we should log it
@nethip Can you please tell the minimum supported platform in windows?

le717 commented on src/thirdparty/uuid.js:
```diff
@@ -0,0 +1,247 @@
+//     uuid.js
```
Is there a minified version we could use here instead or will it automatically be minified when a release is build?

prafulVaishnav replied:
Yes, it will get minified on build.

peterflynn commented:
@prafulVaishnav One high-level comment... this should really be cleanly separated out as a 'core extension' -- i.e. a folder under `src/extensions/default`.  There aren't really any changes to core code needed for this to work.

peterflynn commented on src/config.json:
```diff
@@ -19,7 +19,9 @@
         "linting.enabled_by_default": true,
-        "build_timestamp": ""
+        "build_timestamp": "",
+        "healthDataServerURL" : "http://10.192.16.236:8080",
```
This should be an HTTPS URL to ensure the user's data isn't sniffed over the wire. I know it's anonymous, aggregate, etc., but people still may not want that stuff visible to everyone on their subnet.

Also, can we get a real domain name instead of a bare IP address?  Bare IP looks kinda sketchy, and might limit our options for migrating servers if we need to later.

prafulVaishnav replied:
@peterflynn At the time of raising the PR, integration of server was not done so I used local machine's ip as placeholder. We are working on switching to https and domain name and will update on client as soon as we are done with it

peterflynn replied:
So just to clarify, we're not going to ship 1.3 until this URL is changed, right?

prafulVaishnav replied:
@peterflynn we will be using https. Will update it here as soon as available.

peterflynn commented on src/config.json:
```diff
@@ -19,7 +19,9 @@
-        "build_timestamp": ""
+        "build_timestamp": "",
+        "healthDataServerURL" : "http://10.192.16.236:8080",
+        "healthDataTestServerURL" : "http://10.192.16.236:8080?test=true"
```
I don't think this should be in the config file.  Unit tests should be able to pass in an override URL programmatically... see UpdateNotification-test or ExtensionManager-tet for examples of other places we've done that.

peterflynn replied:
Or for example, instead of passing a boolean `testEnvironment` flag in the URL params during unit tests, you could pass a string with the URL to use.

prafulVaishnav replied:
@peterflynn @abose We will not be needing testServerURL now. We have infrastructure to test it, so not required here.

peterflynn commented on src/healthData/HealthDataManager.js:
```diff
@@ -0,0 +1,180 @@
+        _checkHealthDataExport();
+    });
+
+    window.addEventListener("offline", function () {
```
Do we know these events work reliably in CEF?  I thought @jasonsanjose tried to use them for something in E4B and they didn't actually fire when they should have...

prafulVaishnav replied:
@peterflynn Yes I have tested. Its working.

peterflynn commented on src/healthData/HealthDataManager.js:
```diff
@@ -0,0 +1,180 @@
+                    result.reject();
+                })
+                .always(function () {
+                    PreferencesManager.setViewState("lastTimeSendHealthData", (new Date()).getTime());
```
I think this should be moved into the `.always()` in `_checkHealthDataExport()`, since that's where all the other code that deals with `"lastTimeSendHealthData"` is located.

peterflynn commented on src/healthData/HealthDataManager.js:
```diff
@@ -0,0 +1,180 @@
+
+        if (isHDTracking) {
+
+            var lastTimeSend = PreferencesManager.getViewState("lastTimeSendHealthData"),
```
Similar to Marcel's comment elsewhere, I think these names would be clearer as `lastTimeSent` and `lastTimeSentData`.

peterflynn commented on src/healthData/HealthDataManager.js:
```diff
@@ -0,0 +1,180 @@
+     * Check if health data is to be send to the server. If user has enable the tracking, health data will be send for every 24 hours.
+     * If 24 hours or more than that has been passed, then send health data to the server
+     */
+    function _checkHealthDataExport() {
```
Nit: "export" isn't the best fit... maybe `_checkHealthDataSend()`?

peterflynn commented on src/healthData/HealthDataManager.js:
```diff
@@ -0,0 +1,180 @@
+    /**
+     * Get the health data which will be send to the server. Initially it is only one time data.
+     */
+    function getHealthData() {
```
As this gets more complex, I think it would be good to split this into a different file: one module manages deciding when to send, and doing the actual network IO, and a separate module manages collecting the data & rolling it up into a JSON blob.  E.g. maybe HealthDataCollector & HealthDataUploader?

prafulVaishnav replied:
@peterflynn I have separate branch on which I am working on collecting user events and many more events. It is much more complex than this and such cases is being handled there. We intentionally branched out this PR to keep it very simple for initial release.

peterflynn commented on src/healthData/HealthDataUtils.js:
```diff
@@ -0,0 +1,112 @@
+     */
+    function getUserExtensionsInRegistry(registryObject, userExtensions) {
+        var userInstalledExtensions = [];
+        Object.keys(userExtensions).forEach(function (extensionId) {
```
You can just do:

```
_.forEach(userExtensions, function (extension, extensionId) {
```

peterflynn commented on src/healthData/HealthDataUtils.js:
```diff
@@ -0,0 +1,112 @@
+                registryExtension = registryObject[extensionId];
+            
+            if (extension && extension.installInfo && registryExtension) {
+                userInstalledExtensions.push({"name" : extension.installInfo.metadata.name, "version" : extension.installInfo.metadata.version});
```
You could just use `extensionId` -- same thing as `extension.installInfo.metadata.name`

peterflynn commented on src/healthData/HealthDataUtils.js:
```diff
@@ -0,0 +1,112 @@
+        if (/Windows/.test(userAgent)) {
+            osVersion = winOSStrings[windowRegex.exec(userAgent)[1]] || osVersion;
+        } else if (/Mac/.test(userAgent)) {
+            osVersion = macRegex.exec(userAgent)[1] || osVersion;
```
The `[1]` here and above could crash if the userAgent ever contains something unexpected and the regexp doesn't match.

peterflynn commented on src/healthData/HealthDataUtils.js:
```diff
@@ -0,0 +1,112 @@
+        var osVersion = userAgent;
+        
+        if (/Windows/.test(userAgent)) {
+            osVersion = winOSStrings[windowRegex.exec(userAgent)[1]] || osVersion;
```
I'm not a huge fan of trying to prettify the version strings here on the client side. It's unnecessary and could make data analysis trickier later.  Examples:

* Let's say we condense all versions of Windows "6.3" into the single result string "Windows 8.1 | Windows Server 2012 R2".  Later, MS releases a Windows 8.1 R2 which has "6.3.1" as its version instead. Now we can't distinguish between the two.
* Let's say a new Windows 11 comes out in the future.  If we don't update Brackets early enough, half the data will have the full userAgent as its version (version not found in `winOSStrings`) and half the data will have "Windows 11" as its version (after we update Brackets, now it's recognized).  We'll need special logic on the server to combine the two buckets.
* Same scenario could play out if a CEF update changes the userAgent format on any of the OSes we support, but we don't realize it until after shipping.

So it feels better to either send the raw userAgent (which there's tons of 3rd party libraries we can use to analyze on the server later) or to at least just send the result of the regexp match without any mapping to something else.

prafulVaishnav replied:
@prksingh What you suggest? Should we send raw user agent and did the calculation while generating report? There is one node module to get os version but I think that will be too much to get os version.

peterflynn commented on src/help/HelpCommandHandlers.js:
```diff
@@ -39,7 +39,8 @@ define(function (require, exports, module) {
         AboutDialogTemplate     = require("text!htmlContent/about-dialog.html"),
-        ContributorsTemplate    = require("text!htmlContent/contributors-list.html");
+        ContributorsTemplate    = require("text!htmlContent/contributors-list.html"),
+        HealthDataNotification  = require("healthData/HealthDataNotification");
```
Nit: move this above the `text!` requires.  (Though actually, this change will go away if we move Health Data into a core extension...)

peterflynn commented on src/nls/root/strings.js:
```diff
@@ -417,6 +417,7 @@ define({
     "CMD_RELEASE_NOTES"                   : "Release Notes",
     "CMD_GET_INVOLVED"                    : "Get Involved",
     "CMD_SHOW_EXTENSIONS_FOLDER"          : "Show Extensions Folder",
+    "CMD_HEALTH_DATA_STATISTICS"          : "HealthData Statistics",
```
We should change this to something like "Health Data Sharing."  @ryanstewart, thoughts?

Firefox calls their prefs section "Data Choices," and the menu item to view your data is "Firefox Health Report"... but our menu item combines those two things.  It both shows the data you'll send in the next ping _and_ lets you opt out.

peterflynn commented on src/extensibility/ExtensionManager.js:
```diff
@@ -99,6 +99,25 @@ define(function (require, exports, module) {
+    /**
+     * Registry object container  
+     */
+    var registryObject = {};
```
I might be misunderstanding, but it seems like adding these two fields are unnecessary.  ExtensionManager already exports its `extensions` map, and once `downloadRegistry()` has finished, you can just iterate that and look for entries where both `installInfo` && `registryInfo` are truthy.  (You can keep the check for `installInfo.locationType === LOCATION_USER` too if desired, though it seems unnecessary to me).

prafulVaishnav replied:
@peterflynn extensions map contains every registry extension + local extensions(default + user) So its good to have separate user extension object.

peterflynn replied:
@prafulVaishnav I'm still not convinced this change was needed.  The existing `extensions` object includes both sets of data with clean separation.  You're just duplicating all that data in two other objects, created with code that is largely duplicated code too.

Is there something that wouldn't work right if `HealthDataUtils.getInstalledExtensions()` just did this?

```
        _.forEach(ExtensionManager.extensions, function (extension, extensionId) {
            if (extension.installInfo && extension.registryInfo) {
                userInstalledExtensions.push({"name" : extensionId, "version" : extension.installInfo.metadata.version});
            }
        });
```

peterflynn commented on src/extensibility/ExtensionManager.js:
```diff
@@ -260,6 +280,18 @@ define(function (require, exports, module) {
                 extensions[id] = {};
             }
+            
+            if (locationType === LOCATION_USER) {
```
This bit of code was injected squarely between the `if (!extensions[id])` above and `extensions[id].installInfo =` below, which should stay together as a pair.  For the reasons given in my previous comment I think this whole piece of added code should just go away... but if you do keep it, please move it down so it's not in the middle of that pair.

peterflynn commented on src/htmlContent/healthdata-notification-dialog.html:
```diff
@@ -0,0 +1,16 @@
+<div class="health-data-notification modal">
```
@ryanstewart I'd be interested in your thoughts on the desired UX for this.  Currently, it's _much_ more heavy-handed than what Firefox does, and they've been sort of our model for this feature...

peterflynn commented on src/healthData/HealthDataNotification.js:
```diff
@@ -0,0 +1,93 @@
+            HealthDataPreview.previewHealthDataFile();
+        } else {
+            showDialogHealthDataNotification();
+        }
```
I wonder if we could just do `HealthDataPreview.previewHealthDataFile()` either way.  It seems like the preview dialog is basically a superset of the notification dialog.  And if we were to streamline the notification more (e.g. a message bar like what Firefox uses), then it really couldn't be used here...

peterflynn commented on src/healthData/HealthDataPreview.js:
```diff
@@ -0,0 +1,75 @@
+
+            $template.on("change", "[data-target]:checkbox", function () {
+                newHDPref = $(this).is(":checked");
+            });
```
@MarcelGerber's same point about the checkbox applies here too

peterflynn commented on src/healthData/HealthDataManager.js:
```diff
@@ -0,0 +1,180 @@
+            $.ajax({
+                url: url,
+                type: "POST",
+                data: jsonData,
```
I don't think this is going to serialize the JSON correctly -- I think you ned to explicitly call `JSON.stringify()`, or else it will be converted into a set of query parameters instead.

prafulVaishnav committed:
Move the implementation to the extensions. Data will be send to the server as soon
as user has launched the Brackets and dialog box prompting for user to opt-in
and opt-out is closed. Rework the comments and defects. Test cases has not been
added yet. Will be adding to the next PR.

prafulVaishnav committed:
Moving lastTimeSentData before making AJAX call to make sure data is not sent
before ONE_DAY as setting it to always() will be asynchronous

prafulVaishnav commented:
@peterflynn @MarcelGerber Hii.. Thanks for the review. I have made the changes. Can you please look into it.

prafulVaishnav committed:
Setting ONE_DAY to correct value. Initially value used for testing was left by mistake.
Removing serverTestURL as it is not needed now.

prafulVaishnav commented:
@peterflynn I will be making changes in the dialog box once I get some suggestions from Ryan..

MarcelGerber commented on src/extensions/default/HealthData/HealthDataManager.js:
```diff
@@ -0,0 +1,170 @@
+
+        if (isHDTracking && notificationDialogShown) {
+
+            var lastTimeSent = PreferencesManager.getViewState("lastTimeSentData"),
```
View states are usually not being looked at by the user, but `lastTimeSentData` isn't specific to Health Data.
I'd go with either `lastTimeSentHealthData` or `healthData.lastTimeSent` (the dot notation is probably not a bad idea for the other prefs, too. You could also rewrite it to use a `getExtensionPrefs()` context).

MarcelGerber commented on src/extensions/default/HealthData/HealthDataNotification.js:
```diff
@@ -0,0 +1,98 @@
+     
+            if (id === "save") {
+                newHDPref = $template.find("[data-target]:checkbox").is(":checked");
+                if (hdPref !== newHDPref) {
```
As @peterflynn mentioned, you don't need that check and thus, you don't need `newHDPref` either.

prafulVaishnav replied:
@MarcelGerber There is lot stuff that goes behind while setting the prefs, so its better to have one line check.

MarcelGerber commented on src/extensions/default/HealthData/HealthDataPreview.js:
```diff
@@ -0,0 +1,77 @@
+    var PreferencesManager      = brackets.getModule("preferences/PreferencesManager"),
+        Strings                 = brackets.getModule("strings"),
+        Dialogs                 = brackets.getModule("widgets/Dialogs"),
+        _                       = brackets.getModule("thirdparty/lodash"),
```
nit: lodash is _usually_ the first to be required. You could also try to sort the modules alphabetically, but keeping the `brackets.getModule()`, `require()`, `require("text!")` order.
That's what many other Brackets modules do at least.

MarcelGerber commented on src/extensions/default/HealthData/HealthDataPreview.js:
```diff
@@ -0,0 +1,77 @@
+        
+        HealthDataManager.getHealthData().done(function (healthDataObject) {
+            var content = JSON.stringify(healthDataObject, null, 4);
+            content = _.escape(content);
```
Instead of escaping here, you can also do that in your template file by changing [`{{{content}}}`](https://github.com/adobe/brackets/pull/10795/files#diff-f15f5118c2710f1bb001d4cb2e4b84c5R7) to `{{content}}`

peterflynn replied:
@MarcelGerber That doesn't work unless it's a `<pre>`, which the template isn't using yet...

MarcelGerber replied:
So you're saying Mustache escaping does only work in a `<pre>` element or did I get that wrong?

Normally, it's supposed to act just like `_.escape()`.

peterflynn replied:
I'm saying if you use `{{{content}}}`, then the `<br />` and `&nbsp;` injected on the next two lines will not work correctly.

MarcelGerber commented on src/extensions/default/HealthData/HealthDataPreview.js:
```diff
@@ -0,0 +1,77 @@
+     
+                if (id === "save") {
+                    newHDPref = $template.find("[data-target]:checkbox").is(":checked");
+                    if (hdPref !== newHDPref) {
```
Same as above.

MarcelGerber commented on src/extensions/default/HealthData/HealthDataPreview.js:
```diff
@@ -0,0 +1,77 @@
+    
+    exports.previewHealthData = previewHealthData;
+                                     
+});
```
Please end all your files with a newline.

MarcelGerber commented on src/extensions/default/HealthData/HealthDataManager.js:
```diff
@@ -0,0 +1,170 @@
+            PreferencesManager.setViewState("UUID", userUuid);
+        }
+
+        oneTimeHealthData.guid = userUuid;
```
Not sure if that's intended, but that occurence of `guid` should probably also be replaced.

MarcelGerber commented:
I'm not sure why you deleted the tests. Extension can have tests (in a `unittests.js` file), too.

prafulVaishnav commented:
@MarcelGerber I will be putting tests either into it or next PR.

prafulVaishnav committed:
Adding unit test cases. Submitting changes suggested in PR review.

prafulVaishnav committed:
Decreasing timeout to 4 sec.

prafulVaishnav committed:
Changing function name.

prafulVaishnav committed:
Removing this preference as it should be set to true by default.

nethip committed:
Updating the comments

Made changes to the comments.

nethip committed:
Updating the comments

Updated the comments.

nethip committed:
Adding newline

Added a newline

nethip committed:
Added a newline

Added a newline

nethip commented:
Thanks @prafulVaishnav

peterflynn commented on src/extensions/default/HealthData/HealthDataManager.js:
```diff
@@ -0,0 +1,170 @@
+        
+        window.clearTimeout(timeoutVar);
+        if (isHDTracking && notificationDialogShown) {
+            var lastTimeSent = PreferencesManager.getViewState("lastTimeSentData"),
```
Sorry, I know this name was my suggestion earlier, but @MarcelGerber raised a good point: this prefs key could too easily collide with other stuff.  Could we change to `"lastTimeSentHealthData"`?

peterflynn commented on src/extensions/default/HealthData/HealthDataManager.js:
```diff
@@ -0,0 +1,170 @@
+        }
+
+        oneTimeHealthData.uuid = userUuid;
+        oneTimeHealthData.snapshotTime = (new Date()).getTime();
```
Nit: this and the two below can be simplified to just: `Date.now()`

prafulVaishnav replied:
@peterflynn Thanks. Just compared the performance. Date.now() is slightly better in performance than constructor 'new Date()'

peterflynn commented on src/extensions/default/HealthData/HealthDataPreview.js:
```diff
@@ -0,0 +1,78 @@
+        PreferencesManager      = brackets.getModule("preferences/PreferencesManager"),
+        Strings                 = brackets.getModule("strings"),
+        Dialogs                 = brackets.getModule("widgets/Dialogs"),
+        
```
Nit: blank line should be removed

peterflynn commented on src/extensions/default/HealthData/main.js:
```diff
@@ -0,0 +1,66 @@
+        HealthDataNotification  = require("HealthDataNotification");
+    
+    var menu            = Menus.getMenu(Menus.AppMenuBar.HELP_MENU),
+        healthDataCmdId = "healthData.healthDataStatistics";
```
Nit: our style for constants like this is all-uppercase, like: `HELP_COMMAND_ID`.

peterflynn commented on src/extensions/default/HealthData/unittests.js:
```diff
@@ -0,0 +1,126 @@
+                });
+
+            });
+        });
```
It would be great to add some stronger tests here in the future:

* Ensure the first data-send is not triggered until 24 hours after first launch (you'd have to parameterize the timeout value so tests don't take 1 day to run, though :-)
* Ensure no data is sent if the opt-out preference is set
* Ensure an extension id that's not in the registry isn't returned by HealthDataUtils

peterflynn replied:
Especially important since this code actually no longer follows that first bullet, which may have been an unintentionally-injected bug :-)  See https://github.com/adobe/brackets/pull/10795/files#r27846440.

prafulVaishnav replied:
@peterflynn I have covered all basic unit test cases like dialogs for preview and notification are being displayed, data is getting sent to the server if it has opt-in and not sent to the server if opt-out. I will definitely add more test cases in future as there will be lot change in future implementation when we will start to track user events and others.

peterflynn commented on src/extensions/default/HealthData/HealthDataManager.js:
```diff
@@ -0,0 +1,170 @@
+            var lastTimeSent = PreferencesManager.getViewState("lastTimeSentData"),
+                currentTime = (new Date()).getTime();
+
+            if (!lastTimeSent || (currentTime >= lastTimeSent + ONE_DAY)) {
```
This logic changed a lot since the original review, and there's a big behavior difference: on first launch, `"lastTimeSentData"` isn't set so we now go into this case and send data _immediately_.  My understanding was we wanted to wait 24 hours before sending the first ping -- both so the user has more time to opt out (which is _critical_ if we decide to change the initial notification UI), and also so the first data ping actually contains useful data (once it contains info like average project size, etc.).

The safest way to do that is to immediately set `"lastTimeSentData"` to the current timestamp (`Date.now()`) if we get here and it's not set yet.  That's similar to what the original code did, before commit 07320528.

If it feels too weird to initially set the pref _as if_ data was sent when in fact we haven't sent any yet, you could change the semantics so the pref becomes `"nextHealthDataSendTime"` -- which I think could actually make this code a little easier to understand anyway.

This code is pretty delicate, so be careful if you decide to fix this a different way: I think I see some pitfalls here where other bugs could easily be introduced (like doing math on undefined, or never sending any data if the app isn't left open for > 24h nonstop).

prafulVaishnav replied:
@peterflynn We want to send data as soon as user launch the Brackets. I think @ryanstewart can give more clarification on that.

peterflynn commented:
@prafulVaishnav Sorry, I didn't realize this was merged already.  I added some additional comments today.  Do you think you could address them in a follow-up PR?  I think a number of these points are important to fix before 1.3 ships.  The UI doesn't seem quite finalized/shippable yet either, so I think we'll need more tweaks here before 1.3 either way...

peterflynn commented on src/extensions/default/HealthData/main.js:
```diff
@@ -0,0 +1,66 @@
+    
+    //handles the command execution for healthData menu item
+    function handleHealthDataStatistics() {
+        HealthDataNotification.handleHealthDataStatistics();
```
Changes in the way the HealthDataNotification is implemented could wind up causing the HealthDataManager module to never load (since it's not explicitly a dependency of main.js here).  I wonder if we should add a stub `require("HealthDataNotification");` here, similar to the ones seen in brackets.js ("modules that self-register") to ensure that HealthDataManager and its extremely important AppInit block will definitely run.

Edit: I see we already have such a require() block down in the unit-test code but again, that's safety as a side-effect -- changes in how our tests are structured could make that go away too.  Seems best to explicitly load that module here given how important it is to make sure it starts its ping timer.

prafulVaishnav replied:
@peterflynn My understanding of including this kind of stub was only when it is not used anywhere else and we explicitly want to register it. Can you please tell any scenario in which HealthDataManager will not load. Have we faced any such thing in past?

MarcelGerber commented on src/extensions/default/HealthData/htmlContent/healthdata-notification-dialog.html:
```diff
@@ -0,0 +1,16 @@
+<div class="health-data-notification modal">
```
As @peterflynn was before (https://github.com/adobe/brackets/pull/10795#discussion_r27637175), I'm a concerned about the UX of a modal dialog too, especially for first-time users.
Right after starting Brackets, they'll see a dialog.

ryanstewart replied:
@MarcelGerber yup, agreed. I sent a note to our legal/privacy team with an alternative proposal that provides a toast notification in the upper right hand corner instead of the modal dialog at first launch.

It's still very up front and semi-in-your-face, but I think it's a little bit more elegant way to introduce the concept.
